File: src/tools/cursor.ts
import { usageFromRecord, type AgentRequest, type AgentResult, type AgentUsage } from "./types.js";
import { spawnAgent } from "./spawn.js";

export function buildCursorArgs(request: AgentRequest): string[] {
  const args = ["-p", "--force", "--yolo", "--approve-mcps", "--trust", "--output-format", "json"];
  args.push("--workspace", request.workdir);
  if (request.model) args.push("--model", request.model);
  if (request.nativeId) args.push("--resume", request.nativeId);
  // Prompt as final argv token; Cursor headless accepts the message on argv.
  args.push(request.prompt);
  return args;
}

export function parseCursorOutput(stdout: string): { text: string; nativeId: string; result: unknown; error?: string; usage?: AgentUsage } {
  const trimmed = stdout.trim();
  if (!trimmed) {
    return { text: "", nativeId: "", result: stdout, error: "cursor agent produced empty JSON output" };
  }
  let payload: Record<string, unknown>;
  try {
    payload = JSON.parse(trimmed) as Record<string, unknown>;
  } catch {
    return { text: "", nativeId: "", result: stdout, error: "cursor agent produced invalid JSON output" };
  }
  const text = typeof payload.result === "string" ? payload.result : "";
  const nativeId = typeof payload.session_id === "string" ? payload.session_id : "";
  const usage = usageFromRecord(payload.usage);
  if (!nativeId) return { text, nativeId: "", result: payload, error: "cursor agent JSON missing session_id" };
  return { text, nativeId, result: payload, ...(usage ? { usage } : {}) };
}

export async function runCursor(request: AgentRequest): Promise<AgentResult> {
  const args = buildCursorArgs(request);
  const capture = await spawnAgent(request.command, args, undefined, request.workdir, request.env, "cursor");
  const parsed = parseCursorOutput(capture.stdout);
  return {
    text: parsed.text,
    nativeId: parsed.nativeId,
    exit: capture.exit,
    result: parsed.result,
    stderr: capture.stderr,
    argv: args,
    stdoutBytes: capture.stdout.length,
    ...(parsed.error ? { error: parsed.error } : {}),
    ...(parsed.usage ? { usage: parsed.usage } : {}),
  };
}

File: src/tools/cursor.rs
use serde_json::Value;

use crate::errors::QcError;
use crate::tools::spawn::spawn_agent;
use crate::tools::types::{AgentRequest, AgentResult, AgentUsage, usage_from_record, utf16_units};

pub fn build_cursor_args(request: &AgentRequest) -> Vec<String> {
    let mut args = vec![
        "-p".into(),
        "--force".into(),
        "--yolo".into(),
        "--approve-mcps".into(),
        "--trust".into(),
        "--output-format".into(),
        "json".into(),
        "--workspace".into(),
        request.workdir.clone(),
    ];
    if let Some(model) = &request.model {
        args.push("--model".into());
        args.push(model.clone());
    }
    if let Some(native) = &request.native_id {
        args.push("--resume".into());
        args.push(native.clone());
    }
    args.push(request.prompt.clone());
    args
}

pub struct ParsedCursor {
    pub text: String,
    pub native_id: String,
    pub result: Value,
    pub error: Option<String>,
    pub usage: Option<AgentUsage>,
}

pub fn parse_cursor_output(stdout: &str) -> ParsedCursor {
    let trimmed = stdout.trim();
    if trimmed.is_empty() {
        return ParsedCursor {
            text: String::new(),
            native_id: String::new(),
            result: Value::String(stdout.to_string()),
            error: Some("cursor agent produced empty JSON output".into()),
            usage: None,
        };
    }
    let Ok(payload) = serde_json::from_str::<Value>(trimmed) else {
        return ParsedCursor {
            text: String::new(),
            native_id: String::new(),
            result: Value::String(stdout.to_string()),
            error: Some("cursor agent produced invalid JSON output".into()),
            usage: None,
        };
    };
    let text = payload
        .get("result")
        .and_then(Value::as_str)
        .unwrap_or("")
        .to_string();
    let native_id = payload
        .get("session_id")
        .and_then(Value::as_str)
        .unwrap_or("")
        .to_string();
    let usage = usage_from_record(payload.get("usage"), None, None);
    if native_id.is_empty() {
        return ParsedCursor {
            text,
            native_id: String::new(),
            result: payload,
            error: Some("cursor agent JSON missing session_id".into()),
            usage,
        };
    }
    ParsedCursor {
        text,
        native_id,
        result: payload,
        error: None,
        usage,
    }
}

pub fn run_cursor(request: &AgentRequest) -> Result<AgentResult, QcError> {
    let args = build_cursor_args(request);
    let capture = spawn_agent(
        &request.command,
        &args,
        None,
        &request.workdir,
        &request.env,
        "cursor",
    )?;
    let parsed = parse_cursor_output(&capture.stdout);
    Ok(AgentResult {
        text: parsed.text,
        native_id: parsed.native_id,
        exit: capture.exit,
        result: parsed.result,
        stderr: capture.stderr,
        argv: args,
        stdout_bytes: utf16_units(&capture.stdout),
        error: parsed.error,
        usage: parsed.usage,
    })
}
